### DISCUSSION TITLE
micromon files in /tmp/

### DISCUSSION THREAD

**Discussion by epkumpu** [nextpyp/.github#86] [Q&A]
NextPYP stores the two files micromon.hostprocessor.pid and micromon.hostprocessor.pipe in /tmp/ by default and these are required for running the main process. We have set up nextPYP on a cluster system (non-root), where it works perfectly fine running on a login node with a single instance. However, if we install a second instance, then running this second instance on the same login node becomes impossible as one installation is using these filenames already. We therefore have to install on another login node. Is it possible to change where these two files are kept for a specific instance? Alternatively, would it be possible to include a serial number for each instance's version of these files?

**Comment by cuchaz** [comment]
nextPYP was originally designed to be the only instance running on a single machine, installed by an administrator with root access. We recently introduced a non-root option of installation to help with running on clusters, but we didn't also re-design the system to allow multiple instances on the same node. To support something like that, we'd have to redesign multiple parts of the software since there are several resources (file paths, port numbers, etc) that weren't designed with contention in mind. Maybe instead of a non-root mode, we need some kind of run-out-of-user-folder mode. We'll talk about it internally and try to come up with something that could work here.

**Reply by epkumpu** [reply]
Thanks! I think this ties into an issue of having nextPYP as one central installation in a cluster environment (at least in the computing environment we have access to) and having different users running jobs from the same instance. I don't know if it's standard policy or not, but having more jobs being run under a single username (the person running the nextPYP instance) will push all runs under that instance further back in priority or hits per-username limits on number of jobs. Being able to somehow masquerade as another user based on the nextPYP username in the SLURM queue would really help in this. cryoSPARC actually has the same issue, but they have some sort of masquerade functionality implemented. Too bad it requires permissions regular users do not have, so we're stuck with the same issue there.

**Reply by JuhaHuiskonen** [reply]
For us the only issue seems to be that files are written to `/tmp` and not `/tmp/username` for example.  I may be wrong, but I think the rest (paths, ports etc) can be defined in the `config.toml` and this is specific for each instance.

**Reply by cuchaz** [reply]
We talked about this in our meeting today. Our current thinking is that if we tackle the user masquerade problem, the other multi-instance problems might go away. User masquerading is tricky to get right (in part because of the permissions issues you mentioned), but we have some ideas and will start experimenting with them soon.
